src/platform_impl/macos/view.rs

```diff
@@ -644,54 +731,64 @@ extern "C" fn key_down(this: &Object, _sel: Sel, event: id) {
         let state_ptr: *mut c_void = *this.get_ivar("winitState");
         let state = &mut *(state_ptr as *mut ViewState);
         let window_id = WindowId(get_window_id(state.ns_window));
+
+        let input_source = current_input_source();
+        if state.input_source != input_source && state.is_ime_enabled() {
+            state.ime_state = ImeState::Disabled;
+            state.input_source = input_source;
+            AppState::queue_event(EventWrapper::StaticEvent(Event::WindowEvent {
+                window_id: WindowId(get_window_id(state.ns_window)),
+                event: WindowEvent::IME(IME::Disabled),
+            }));
+        }
+        let was_in_preedit = state.ime_state == ImeState::Preedit;
+
         let characters = get_characters(event, false);
+        state.forward_key_to_app = false;
+
+        // The `interpretKeyEvents` function might call
+        // `setMarkedText`, `insertText`, and `doCommandBySelector`.
+        // It's important that we call this before queuing the KeyboardInput, because
+        // we must send the `KeyboardInput` event during IME if it triggered
+        // `doCommandBySelector`. (doCommandBySelector means that the keyboard input
+        // is not handled by IME and should be handled by the application)
+        if state.app_allowed_ime {
+            let events_for_nsview: id = msg_send![class!(NSArray), arrayWithObject: event];
+            let _: () = msg_send![this, interpretKeyEvents: events_for_nsview];
+        }
 
-        state.raw_characters = Some(characters.clone());
+        let now_in_preedit = state.ime_state == ImeState::Preedit;
 
         let scancode = get_scancode(event) as u32;
         let virtual_keycode = retrieve_keycode(event);
-
-        let is_repeat: BOOL = msg_send![event, isARepeat];
-
+        // let is_repeat: BOOL = msg_send![event, isARepeat];
         update_potentially_stale_modifiers(state, event);
 
-        #[allow(deprecated)]
-        let window_event = Event::WindowEvent {
-            window_id,
-            event: WindowEvent::KeyboardInput {
-                device_id: DEVICE_ID,
-                input: KeyboardInput {
-                    state: ElementState::Pressed,
-                    scancode,
-                    virtual_keycode,
-                    modifiers: event_mods(event),
+        let preedit_related = was_in_preedit || now_in_preedit;
+        if !preedit_related || state.forward_key_to_app || !state.app_allowed_ime {
```

Comment:
I _think_ accessing `state.forward_key_to_app` is pretty bad here without a mutex/atomics:
`state: &mut ViewState`, and since it is not passed down to `interpretKeyEvents`, Rust is free to reorder or even elide the access (since it "knows" no-one could have modified `forward_key_to_app` between where it was set to `true`, and here).

Reply:
Maybe we can get rid of the `forward_key_to_app` variable altogether, and instead do the work inside `doCommandBySelector:`? Also, I can't seem to find a place where `doCommandBySelector:` is triggered while IME is active, can you give me some examples of this?

Reply:
Very well spotted! I believe the easiest fix for us is using a [compiler_fence](https://doc.rust-lang.org/std/sync/atomic/fn.compiler_fence.html) after the `interpretKeyEvents`

> can you give me some examples of this?

With the Pinyin input active, press N, then press backspace, then press the up arrow. The up arrow will trigger `doCommandBySelector`.